core/types/transaction_marshalling.go

```diff
@@ -150,6 +156,11 @@ func (tx *Transaction) MarshalJSON() ([]byte, error) {
 		enc.V = (*hexutil.Big)(itx.V)
 		enc.R = (*hexutil.Big)(itx.R)
 		enc.S = (*hexutil.Big)(itx.S)
+		return json.Marshal(
+			&txJSONCelo{
+				txJSON:      enc,
+				FeeCurrency: enc.FeeCurrency,
+			})
```

Comment:
I don't think there is any need for `txJSONCelo` you can just do the following here:
```suggestion
		enc.FeeCurrency = tx.FeeCurrency()
```

Reply:
But then we would not have different behavior of marshaling the `feeCurrency` depending on which transaction type there is. We want the feeCurrency field to be excluded when we encode a non-CIP64/66 transaction and we want it to be explicitly `"feeCurrency": "null"` when a CIP64 transaction without fee-currency is encoded.
For CIP64 transactions it is possible to have a null feeCurrency, right? The [specs](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0064.md) don't say wether the field is optional or not. CIP-66 does not allow an optional feeCurrency.

We could also make it `omitempty` for all variants and never allow for encoding `"feeCurrency": "null"`

Reply:
Hey @ezdac, so I looked a bit deeper into this and it seems to me that the txJSON struct is just used as an intermediate decoding step that `Transaction.MarshalJSON` & `Transaction.UnmarshalJSON` rely on. So it doesn't really matter if `FeeCurrency` is marshaled as `"feeCurrency": "null"` or not at all, because in transaction marshaling/unmarshaling whether or not to include that field is determined by the transaction type that controls the internal switch statement.

So I still think just using `omitempty` should be sufficient.

Reply:
Yes, for roundtrip encode - decode inside the op-geth it doesn't matter. And also nothing behaves differently for the internal Transaction type upon decoding when comparing `"feeCurrency": "null"` vs. it being left out.

We were just thinking about wether other clients interacting with the API would expect an explicit "null" for CIP64 types in terms of it being a semantic difference vs. the omission of the field. But I think this is splitting hairs and I don't see a very strong semantic reason for this, especially since CIP64 might get deprecated at some point. 

So I'll go with the simpler `omitempty` for all approach.

Reply:
Ahh but note the JSON encoding here is not what is provided over the RPC API.  The RPC json serialization is handled with [this](https://github.com/celo-org/op-geth/blob/celo4/internal/ethapi/api.go#L1545) struct.

Reply:
Ohh okay, I should have already gotten that from your earlier comment.

Yeah makes sense - it is a little confusing that the `api_test.go` does not test the actual JSON encoding but the intermediary one. So now `omitempty` all is definitely the way to go.

Reply:
Yep, the encoding/decoding is highly confusing! There's also [SendTxArgs](https://github.com/celo-org/op-geth/blob/b510efe282ea4e8bfa11cfc84917644f75d0bdeb/signer/core/apitypes/types.go#L82) and [TransactionArgs](https://github.com/celo-org/op-geth/blob/b510efe282ea4e8bfa11cfc84917644f75d0bdeb/internal/ethapi/transaction_args.go#L39).

So 4 different ways and structs to JSON serialize a transaction! There could even be more, but those are the ones I'm currently aware of.

Reply:
Oh, good to know, that's great - I wasn't aware of the `SendTxArgs`, but this might also be relevant for the FeeCurrency changes!